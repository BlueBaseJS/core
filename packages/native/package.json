{
  "name": "bluerain-boilerplate-native",
  "version": "0.3.0",
  "description": "Hello BlueRain!",
  "author": "BlueEast Team <code@blueeast.com>",
  "private": true,
  "main": "node_modules/expo/AppEntry.js",
  "scripts": {
    "start": "exp start"
  },
  "dependencies": {
    "@blueeast/bluerain-os": "git://github.com/BlueEastCode/bluerain-os.git#feature/custom_platform",
    "expo": "19.0.1",
    "react": "16.0.0-alpha.12",
    "react-dom": "^16.0.0-alpha.12",
    "react-native": "0.46.4",
    "react-native-windows": "^0.33.0",
    "reactxp": "^0.42.0-rc.25",
    "reactxp-navigation": "1.0.5",
    "reactxp-video": "^0.2.0"
  },
  "devDependencies": {
    "@types/node": "^7.0.12",
    "@types/react": "15.0.39",
    "@types/react-native": "0.46.7",
    "@types/webpack": "^2.2.14",
    "source-map-loader": "^0.1.6"
  }
}
