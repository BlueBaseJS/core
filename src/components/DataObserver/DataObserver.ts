import { DataObserverProps } from '@bluebase/components';
import React from 'react';
import isboolean from 'lodash.isboolean';
import isnil from 'lodash.isnil';

/**
 * # ⚡️ DataObserver
 *
 * Observes data to check if it is data is loading, loaded or empty. The resulting flags
 * are passed on to the children function. These flags may be used to show different UIs,
 * i.e. loading state, empty state, etc.
 *
 * ## Usage
 * ```jsx
 * <DataObserver>
 *  <Text>{data}</Text>
 * </DataObserver>
 * ```
 */
export class DataObserver extends React.PureComponent<DataObserverProps> {
	public static defaultProps: Partial<DataObserverProps> = {
		loading: false,

		isEmpty: props => {
			// If its null or undefined
			if (isnil(props.data)) {
				return true;
			}

			// If its an empty array
			if (Array.isArray(props.data) && props.data.length === 0) {
				return true;
			}

			return false;
		},
		isLoading: props => (!isnil(props.loading) && isboolean(props.loading) ? props.loading : false),
	};

	render() {
		const { children } = this.props;

		if (typeof children === 'function') {
			return (children as any)({
				data: this.props.data,
				empty: this.props.isEmpty ? this.props.isEmpty(this.props) : false,
				loading: this.props.isLoading ? this.props.isLoading(this.props) : false,
			});
		}

		return children;
	}
}
