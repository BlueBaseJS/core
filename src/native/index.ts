export * from './ActivityIndicator';
export * from './Button';
export * from './Image';
export * from './Text';
export * from './View';
