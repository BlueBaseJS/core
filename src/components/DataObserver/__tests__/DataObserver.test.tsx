import * as Native from '../../../native';
import { Text, View } from 'react-native';
import { BlueBaseApp } from '../../../';
import { DataObserver } from '../DataObserver';
import React from 'react';
import TestRenderer from 'react-test-renderer';

describe('DataObserver', () => {

	test(`Snapshot DataObserver component with Text`, (done) => {
		const component = TestRenderer.create(
			<BlueBaseApp>
				<DataObserver>
					<Native.Text>hello</Native.Text>

				</DataObserver>
			</BlueBaseApp>
		);
		setTimeout(() => {
			const tree = component.toJSON();
			expect(tree).toMatchSnapshot();
			done();
		});
	});

	test(`should generate states from custom listener functions`, (done) => {
		const component = TestRenderer.create(
			<BlueBaseApp>
				<DataObserver data="foo" isEmpty={() => false} isLoading={() => true}>
				{({ data, empty, loading }: { data: string, empty: boolean, loading: boolean }) => {
					return (
						<View>
							<Text>{data}</Text>
							<Text>{String(empty)}</Text>
							<Text>{String(loading)}</Text>
						</View>
					);
				}}
				</DataObserver>
			</BlueBaseApp>
		);
		setTimeout(() => {
			const tree = component.toJSON();
			expect((tree as any).children[0].children.join()).toBe('foo');
			expect((tree as any).children[1].children.join()).toBe('false');
			expect((tree as any).children[2].children.join()).toBe('true');
			expect(tree).toMatchSnapshot();
			done();
		});
	});

	test(`should set loading to true`, (done) => {
		const component = TestRenderer.create(
			<BlueBaseApp>
				<DataObserver data="foo" loading={true}>
				{({ data, empty, loading }: { data: string, empty: boolean, loading: boolean }) => {
					return (
						<View>
							<Text>{data}</Text>
							<Text>{String(empty)}</Text>
							<Text>{String(loading)}</Text>
						</View>
					);
				}}
				</DataObserver>
			</BlueBaseApp>
		);
		setTimeout(() => {
			const tree = component.toJSON();
			expect((tree as any).children[0].children.join()).toBe('foo');
			expect((tree as any).children[1].children.join()).toBe('false');
			expect((tree as any).children[2].children.join()).toBe('true');
			expect(tree).toMatchSnapshot();
			done();
		});
	});

	test(`should set empty to true if data if an empty array`, (done) => {
		const component = TestRenderer.create(
			<BlueBaseApp>
				<DataObserver data={[]} loading={true}>
				{({ data, empty, loading }: { data: string, empty: boolean, loading: boolean }) => {
					return (
						<View>
							<Text>{data}</Text>
							<Text>{String(empty)}</Text>
							<Text>{String(loading)}</Text>
						</View>
					);
				}}
				</DataObserver>
			</BlueBaseApp>
		);
		setTimeout(() => {
			const tree = component.toJSON();
			expect((tree as any).children[0].children).toBe(null);
			expect((tree as any).children[1].children.join()).toBe('true');
			expect((tree as any).children[2].children.join()).toBe('true');
			expect(tree).toMatchSnapshot();
			done();
		});
	});

	test(`should set empty to false if data if an array`, (done) => {
		const component = TestRenderer.create(
			<BlueBaseApp>
				<DataObserver data={['a']} loading={true}>
				{({ data, empty, loading }: { data: string, empty: boolean, loading: boolean }) => {
					return (
						<View>
							<Text>{(data as any).join()}</Text>
							<Text>{String(empty)}</Text>
							<Text>{String(loading)}</Text>
						</View>
					);
				}}
				</DataObserver>
			</BlueBaseApp>
		);
		setTimeout(() => {
			const tree = component.toJSON();
			expect((tree as any).children[0].children.join()).toBe('a');
			expect((tree as any).children[1].children.join()).toBe('false');
			expect((tree as any).children[2].children.join()).toBe('true');
			expect(tree).toMatchSnapshot();
			done();
		});
	});

	test(`should set empty and loading to false when there are no listener functions`, (done) => {
		const component = TestRenderer.create(
			<BlueBaseApp>
				<DataObserver data="a" isEmpty={null as any} isLoading={null as any}>
				{({ data, empty, loading }: { data: string, empty: boolean, loading: boolean }) => {
					return (
						<View>
							<Text>{data}</Text>
							<Text>{String(empty)}</Text>
							<Text>{String(loading)}</Text>
						</View>
					);
				}}
				</DataObserver>
			</BlueBaseApp>
		);
		setTimeout(() => {
			const tree = component.toJSON();
			expect((tree as any).children[0].children.join()).toBe('a');
			expect((tree as any).children[1].children.join()).toBe('false');
			expect((tree as any).children[2].children.join()).toBe('false');
			expect(tree).toMatchSnapshot();
			done();
		});
	});

	test(`should set loading to false, if loading prop is not a boolean`, (done) => {
		const component = TestRenderer.create(
			<BlueBaseApp>
				<DataObserver data="a" loading={'true' as any}>
				{({ data, empty, loading }: { data: string, empty: boolean, loading: boolean }) => {
					return (
						<View>
							<Text>{data}</Text>
							<Text>{String(empty)}</Text>
							<Text>{String(loading)}</Text>
						</View>
					);
				}}
				</DataObserver>
			</BlueBaseApp>
		);
		setTimeout(() => {
			const tree = component.toJSON();
			expect((tree as any).children[0].children.join()).toBe('a');
			expect((tree as any).children[1].children.join()).toBe('false');
			expect((tree as any).children[2].children.join()).toBe('false');
			expect(tree).toMatchSnapshot();
			done();
		});
	});

});
