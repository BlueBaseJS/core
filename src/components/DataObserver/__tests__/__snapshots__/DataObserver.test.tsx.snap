// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataObserver Snapshot DataObserver component with Text 1`] = `
<View
  styles={Object {}}
>
  <Text
    style={
      Array [
        undefined,
        undefined,
      ]
    }
  >
    hello
  </Text>
</View>
`;

exports[`DataObserver should generate states from custom listener functions 1`] = `
<View
  styles={Object {}}
>
  <View>
    <Text>
      foo
    </Text>
    <Text>
      false
    </Text>
    <Text>
      true
    </Text>
  </View>
</View>
`;

exports[`DataObserver should set empty and loading to false when there are no listener functions 1`] = `
<View
  styles={Object {}}
>
  <View>
    <Text>
      a
    </Text>
    <Text>
      false
    </Text>
    <Text>
      false
    </Text>
  </View>
</View>
`;

exports[`DataObserver should set empty to false if data if an array 1`] = `
<View
  styles={Object {}}
>
  <View>
    <Text>
      a
    </Text>
    <Text>
      false
    </Text>
    <Text>
      true
    </Text>
  </View>
</View>
`;

exports[`DataObserver should set empty to true if data if an empty array 1`] = `
<View
  styles={Object {}}
>
  <View>
    <Text />
    <Text>
      true
    </Text>
    <Text>
      true
    </Text>
  </View>
</View>
`;

exports[`DataObserver should set loading to false, if loading prop is not a boolean 1`] = `
<View
  styles={Object {}}
>
  <View>
    <Text>
      a
    </Text>
    <Text>
      false
    </Text>
    <Text>
      false
    </Text>
  </View>
</View>
`;

exports[`DataObserver should set loading to true 1`] = `
<View
  styles={Object {}}
>
  <View>
    <Text>
      foo
    </Text>
    <Text>
      false
    </Text>
    <Text>
      true
    </Text>
  </View>
</View>
`;
